marshal/consumer.go

```diff
@@ -375,6 +394,35 @@ func (c *Consumer) claimTopics() {
 		}
 	}
 
+	// let's compare the new topic claims vs old ones. Upon change, we should trigger an update
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	changed := false
+	// check for missing topic claims
+	for topic := range c.claimedTopics {
+		if !latestClaims[topic] {
+			changed = true
+			break
+		}
+	}
+	// check for new topic claims
+	if !changed && len(latestClaims) != len(c.claimedTopics) {
+		changed = true
+	}
+
+	if changed {
+		c.lock.RUnlock()
```

Comment:
I know it can make it harder to reason about in some cases, but I think that serial locking leads to its own set of deadlocks and hard to reason about problems. Honestly @DrTall convinced me of this logic, he might be able to elaborate on it better than I.